### ISSUE TITLE
Configurable intermediate chain (per authority/provisioner) for appending to returned certificate bundle

### ISSUE THREAD

**Issue by viralpoetry** [smallstep/certificates#244]
### What would you like to be added   

Hi, 
We would like to leverage smallstep certificates as issuing CA (3rd in the hierarchy) but we have problem on how to setup CA to serve the chain correctly (I think it's not possible at all?).

### Why this is needed  
As an organization, we are using 3 tier hierarchy where Root and Intermedate CAs are not under our control and we do not have access to a corresponding private keys.

Hierarchy should look like:
```
Root CA (HSM protected) -> Intermediate CA (HSM protected) -> Smallstep CA (issuer, software only, serving the whole chain to the ACME clients)
```

I tried to modify config file to serve our need, but I can't make it work with this kind of chain.
There is #87 but I can't decide if it's duplicate or other use case as I don't want to manage CAs in hierarchy with `smallstep certificates` at all.

Thank you for making great software,
Peter Gasper

**Comment by mkkeffeler** [comment]
We are trying to do a similar thing, where we have our own root and we want to sign a subordinate CA, then have smallstep use that as its root CA, for which it generates a subordinate off of that. in this way we better protect our root CA. I don't imagine this is the first time this is coming up, but I am having the same issue described here right now and would very much appreciate assistance on if this is possible.

**Comment by dopey** [comment]
Heyo, y'all are right - this is not the first time this has come up. Am I right to assume that the main issue is the max path len on the certificates?

**Comment by mkkeffeler** [comment]
Not sure, Let me walk through what I did briefly.

Generated a root CA, signed a subordinate CA with my root using the v3_ca extension, then gave step ca init the subordinate CA key and cert, to which it then generated a leaf certificate off of that, but now I get this error when I try to claim a token from that CA 

SC-MKEFFEL-OSX:steppoc5 mkeffele$ step ca token subsigned.autozone.com -ca-url=https://localhost:8443 --root ../../steppoc6/cacert.pem 
client GET https://localhost:8443/provisioners?limit=100 failed: Get "https://localhost:8443/provisioners?limit=100": x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "rootca.autozone.com")


Where rootca.autozone.com is the root CA that signed the intermediate i used in step ca init.

**Comment by maraino** [comment]
@mkkeffeler Not 100% related to this issue, but we currently support Google's KMS too, and we'll be adding support for other platforms.

**Comment by dopey** [comment]
@mkkeffeler Want to hop in our [gitter](https://gitter.im/smallstep/community)? It can be a bit easier to help debug in real time.

If not, no worries, I can try and help async as well.

**Comment by viralpoetry** [comment]
@dopey thanks for reaching out so fast. I would say that the problem is similar, but not the same as described by @mkkeffeler , as I can't let `certificates` to just use already existing (issued 3rd party) CA. In my case, I have a signed issuing CA certificate (3rd in a hierarchy) & private key, and part of a problem is on how to serve the whole chain if I can't setup smallstep with just the issuing cert, key alone.

**Comment by dopey** [comment]
@viralpoetry I see. That's interesting, so you'd like `certificates` to return certs that have your full chain. Unfortunately, that's not possible at the moment. 

From what I understand, here is your chain: 

1. root
2. inter1
3. inter2 (this is the root for step-ca)
4. leaf

`certificates` returns 4+3. And what you want is 4+3+2+1. Which means that you will need to manually append 2/1 to every cert generated by `step-ca` for the time being.

The other option would  be to add inter2 to the truststore, allowing your servers/clients to use it as a root.

**Comment by mkkeffeler** [comment]
Max helped me resolve the issue I described. I had incorrectly configured my step CA and thus was having issues. Once I had correctly gotten a key and crt for the subordinate I was able to configure what I described without issue. Thanks to this team!

**Comment by dopey** [comment]
@viralpoetry Let me know if my last comment is a correct characterization of your use case. If so, maybe we can change the title of this issue (or create a new issue with a more specific title).

For the time being I'll mark this issue as `needs more info` and maybe close the issue in a few weeks if I don't hear back.

If anyone else in the community has a similar use case, feel free to open the "more specific issue" (adding intermediate certificates to the returned chain) that I referenced above.

**Comment by viralpoetry** [comment]
@dopey Hi, your last comment described it correctly:
https://github.com/smallstep/certificates/issues/244#issuecomment-623738394

Please, do change the title for a more appropriate name. I would be interested to make it work in the future. However I had to solve it with other software for now

**Comment by dopey** [comment]
Awesome! appreciate the feedback.

**Comment by mkkeffeler** [comment]
i think that there is still a reason for this. in my example, my issue was i wasn't seeing that smallstep was correctly utilizing the CA i gave it (which was an intermediate to my root). we did resolve that, but now when we try to use that cert the full chain has to still be appended to manually (with the intermediate that step has been given as its root, and the true root ca). any thoughts/workarounds to this?

**Comment by christiaangoossens** [comment]
I also have this problem, external Root CA and Intermediate CA. That intermediate CA signed a CA that is now used as the Step CA (like the instructions given). The HTTPS endpoint for step-ca now returns its leaf certificate + that Step CA, but because the Intermediate CA is missing in the HTTPS response, I get errors. Any method to add a bundle to the HTTPS server for step-ca?

**Comment by mkkeffeler** [comment]
with templates now released, (AND AMAZING, might I add, already using). Is there something we can fix this with? short story, we have a root CA that smallstep doesn't know about (it signed the intermediate that smallstep is using as its root), that we want to deliver as part of the chain of trust when users get a certificate from us.

**Comment by maraino** [comment]
Templates won't for supporting multiple intermediates, this seems to be a duplicate of (https://github.com/smallstep/certificates/issues/87)

It's not a P0 priority right now, but it's something we will add. The support must be added to both `step` and `step-ca`, and the basic building blocks are already in `step-ca`. We always appreciate contributions to the project.

**Comment by dwchurch** [comment]
This is an issue for me and my people

**Comment by razron** [comment]
Got similar problems as described by @mkkeffeler  We are having to manually append the additional certificates or use additional script logic to add them. Not all clients make it easy to do the additional scripting and adding things manually defeats automatic renewals. So another me too.

**Comment by maraino** [comment]
@mkkeffeler, @dwchurch, @razron Can you please add a 👍 to https://github.com/smallstep/certificates/issues/87

**Comment by mkkeffeler** [comment]
thee body of the issue

**Comment by maraino** [comment]
Fixed by #440. That PR adds support of a certificate chain of multiple intermediates in the full authority, not a per-provisioner one. If a per provisioner one is required a new issue should be opened.
